### ISSUE TITLE
Crash when resolving calls to variadic template functions

### ISSUE THREAD

**Issue by peckto** [Fraunhofer-AISEC/cpg#767]
# Stack trace

```
java.util.concurrent.ExecutionException: java.lang.NullPointerException
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:395)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1999)
	at de.fraunhofer.aisec.cpg_vis_neo4j.Application.call(Application.kt:402)
	at de.fraunhofer.aisec.cpg_vis_neo4j.Application.call(Application.kt:70)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1953)
	at picocli.CommandLine.access$1300(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2352)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2346)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2311)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
	at picocli.CommandLine.execute(CommandLine.java:2078)
	at de.fraunhofer.aisec.cpg_vis_neo4j.ApplicationKt.main(Application.kt:442)
Caused by: java.lang.NullPointerException
	at de.fraunhofer.aisec.cpg.passes.CallResolver.getCallSignature(CallResolver.java:733)
	at de.fraunhofer.aisec.cpg.passes.CallResolver.handleTemplateFunctionCalls(CallResolver.java:557)
	at de.fraunhofer.aisec.cpg.passes.CallResolver.handleNormalCallCXX(CallResolver.java:1062)
	at de.fraunhofer.aisec.cpg.passes.CallResolver.handleNormalCalls(CallResolver.java:1024)
	at de.fraunhofer.aisec.cpg.passes.CallResolver.handleCallExpression(CallResolver.java:299)
	at de.fraunhofer.aisec.cpg.passes.CallResolver.resolve(CallResolver.java:260)
	at de.fraunhofer.aisec.cpg.helpers.SubgraphWalker$ScopedWalker.lambda$registerHandler$0(SubgraphWalker.java:408)
	at de.fraunhofer.aisec.cpg.helpers.SubgraphWalker$ScopedWalker.handleNode(SubgraphWalker.java:445)
	at de.fraunhofer.aisec.cpg.helpers.SubgraphWalker$ScopedWalker.lambda$iterate$1(SubgraphWalker.java:418)
	at de.fraunhofer.aisec.cpg.helpers.SubgraphWalker$IterativeGraphWalker.lambda$iterate$1(SubgraphWalker.java:333)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at de.fraunhofer.aisec.cpg.helpers.SubgraphWalker$IterativeGraphWalker.iterate(SubgraphWalker.java:333)
	at de.fraunhofer.aisec.cpg.helpers.SubgraphWalker$ScopedWalker.iterate(SubgraphWalker.java:420)
	at de.fraunhofer.aisec.cpg.passes.CallResolver.accept(CallResolver.java:102)
	at de.fraunhofer.aisec.cpg.passes.CallResolver.accept(CallResolver.java:63)
	at de.fraunhofer.aisec.cpg.TranslationManager.analyze$lambda-2(TranslationManager.kt:103)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1692)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
```

# Unexpected behavior

When resolving calls to variadic template functions, the cpg crashes under certain circumstances.
If the respective function does exist, the calls is resolved correctly.
If the respective function name does not exist, the function is inferred.
If a function with the same name, but different arguments exist, the cpg crashes!

I'm not quite sure about the root cause of this crash.
As far as I understand, fist possible template function candidates are resolved:
```kt
    fun resolveFunctionTemplateDeclaration(
        call: CallExpression,
        scope: Scope? = currentScope
    ): List<FunctionTemplateDeclaration> {
        return resolve(scope, true) { c: FunctionTemplateDeclaration -> c.name == call.name }
    }
```

In this first step, only the function name is taken into account.
In case a template function with same name (but possible different parameters) exists, this function is returned.
When resolving/mapping the function parameters in `getCallSignature` it does not find the parameter and crashes with a `NullPointerException`.

This issue might be related to #763 

# Code of interest

```cpp
#include <iostream>
#include <typeinfo>

/*template <typename TT>
TT sum(TT arg)
{
    return arg;
}*/


template <typename T, typename... Args>
T sum(T first, Args... args)
{
    sum(args...);
    return first;
}

int main(int argc, char* argv[])
{
    auto result = sum(true, 0, 4.6, 3.14f);
    std::cout << typeid(result).name() << ": " << result << '\n';
    return 0;
}
```

# Extensions to the library, e.g. added or changed LanguageFrontends or Passes

No changes to the library

# Changes to the configuration

`cpg-neo4j  main.cpp`

**Comment by shenniger** [comment]
This has been observed in this real-world code file: https://github.com/google/gemmlowp/blob/master/test/test.cc

There, a template function has two overloads: one with two template arguments, the second with only one. Upon calling the function with only one template argument, the CPG crashes (although the code is well-formed and compiles in GCC):
```c++
template <typename A, B c> void fun1() {}
template <typename A> void fun1() {}
void fun2() { fun1<SomeType>(); }
```

The following equivalent code example does not lead to the crash and leads to a correct graph (afaict):
```c++
template <typename A> void fun1() {}
template <typename A, B c> void fun1() {}
void fun2() { fun1<SomeType>(); }
```

**Comment by shenniger** [comment]
See #1177 on how to fully solve this. As a hotfix to fix only the exception and keep analyzing code on a best-effort basis, this seems to work:
```
--- a/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/CXXCallResolverHelper.kt
+++ b/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/CXXCallResolverHelper.kt
@@ -369,7 +369,9 @@ fun applyTemplateInstantiation(
     // Add DFG edges from the instantiation Expression to the ParamVariableDeclaration in the
     // Template.
     for ((declaration) in initializationSignature) {
-        if (declaration is ParamVariableDeclaration) {
+        if (
+            declaration is ParamVariableDeclaration && initializationSignature[declaration] != null
+        ) {
             declaration.addPrevDFG(initializationSignature[declaration]!!)
             initializationSignature[declaration]!!.addNextDFG(declaration)
         }

```

**Comment by KuechA** [comment]
The NPE from the provided example is fixed in #1181 (with a solution similar to the proposal of @shenniger).
